gpcontrib/gpcloud/src/gpcloud.cpp

```diff
@@ -274,7 +274,7 @@ Datum s3_import(PG_FUNCTION_ARGS) {
 
     /* last call. destroy reader */
     if (EXTPROTOCOL_IS_LAST_CALL(fcinfo)) {
-        destroyGpcloudResHandle(resHandle);
+        destroyGpcloudResHandle(resHandle, ERROR);
```

Comment:
Do we need to throw an error here? If we're in this condition, we'll return anyway. Plus, `resHandle` is freed in `destroyGpcloudResHandle`, so is it possible for it to leak if we throw here? It seems to me that having this callback called on resowner destruction is not right and we might consider rewriting it so that it would be called on transaction end, as Vasiliy mentioned

Reply:
We should somehow signal to the user that the cleanup resulted in an error, because it doesn't actually send any data to S3 in that case. s3_import and s3_export are regular functions so they won't be called during the resowner destruction, and so everything should be cleaned up properly once resowner gets destroyed. The only function here that can be called on resowner destruction is `gpcloudAbortCallback` and now this will only print a warning.

Reply:
I think that the solution is a bit controversial, because we can get the same error text with different statuses. I understand that it's the most simple solution which doesn't require you to rewrite too much code, so I'll approve, but I'll leave this conversation open, so that the next reviewer would be able to leave his opinion